### PR TITLE
NUTCH-2809: Upgrade any23 plugin dependency

### PR DESCRIPTION
**Pull request by balashashanka**
Upgraded the library to any23-2.3. The tests pass for the plugin.

### CODE REVIEW

**Comment by lewismc** [comment]
I'll check this one out shortly. Good work @balashashanka :)

**Comment by lewismc** [comment]
@balashashanka 

```
rm -r ~/.m2/repository
ant clean test
...
init:
    [mkdir] Created dir: /Users/lmcgibbn/Downloads/nutch/build/any23/classes
    [mkdir] Created dir: /Users/lmcgibbn/Downloads/nutch/build/any23/test/lib
    [mkdir] Created dir: /Users/lmcgibbn/Downloads/nutch/build/plugins/any23

init-plugin:

deps-jar:

clean-lib:

resolve-default:
[ivy:resolve] :: loading settings :: file = /Users/lmcgibbn/Downloads/nutch/ivy/ivysettings.xml
[ivy:resolve]
[ivy:resolve] :: problems summary ::
[ivy:resolve] :::: WARNINGS
[ivy:resolve] 		[FAILED     ] javax.ws.rs#javax.ws.rs-api;2.1.1!javax.ws.rs-api.${packaging.type}:  (0ms)
[ivy:resolve] 	==== local: tried
[ivy:resolve] 	  /Users/lmcgibbn/.ivy2/local/javax.ws.rs/javax.ws.rs-api/2.1.1/${packaging.type}s/javax.ws.rs-api.${packaging.type}
[ivy:resolve] 	==== maven2: tried
[ivy:resolve] 	  https://repo1.maven.org/maven2/javax/ws/rs/javax.ws.rs-api/2.1.1/javax.ws.rs-api-2.1.1.${packaging.type}
[ivy:resolve] 	==== apache-snapshot: tried
[ivy:resolve] 	  https://repository.apache.org/content/repositories/snapshots/javax/ws/rs/javax.ws.rs-api/2.1.1/javax.ws.rs-api-2.1.1.${packaging.type}
[ivy:resolve] 	==== sonatype: tried
[ivy:resolve] 	  https://oss.sonatype.org/content/repositories/releases/javax/ws/rs/javax.ws.rs-api/2.1.1/javax.ws.rs-api-2.1.1.${packaging.type}
[ivy:resolve] 		::::::::::::::::::::::::::::::::::::::::::::::
[ivy:resolve] 		::              FAILED DOWNLOADS            ::
[ivy:resolve] 		:: ^ see resolution messages for details  ^ ::
[ivy:resolve] 		::::::::::::::::::::::::::::::::::::::::::::::
[ivy:resolve] 		:: javax.ws.rs#javax.ws.rs-api;2.1.1!javax.ws.rs-api.${packaging.type}
[ivy:resolve] 		::::::::::::::::::::::::::::::::::::::::::::::
[ivy:resolve]
[ivy:resolve] :: USE VERBOSE OR DEBUG MESSAGE LEVEL FOR MORE DETAILS

BUILD FAILED
/Users/lmcgibbn/Downloads/nutch/build.xml:116: The following error occurred while executing this line:
/Users/lmcgibbn/Downloads/nutch/src/plugin/build.xml:38: The following error occurred while executing this line:
/Users/lmcgibbn/Downloads/nutch/src/plugin/build-plugin.xml:229: impossible to resolve dependencies:
	resolve failed - see output for details

Total time: 21 seconds
```

**Comment by balashashanka** [comment]
Hi @lewismc, this is a well known issue in javax.ws.rs-api. See [here](https://mvnrepository.com/artifact/javax.ws.rs/javax.ws.rs-api/2.1.1), files are available in ${packaging.type}. So this can be overcome using -Dpackaging.type=jar (something similar) while doing command line and that should set that. I initially thought it was my system then found this. Try and check if this fixes it.

File: src/plugin/any23/ivy.xml [hunk 1]
```diff
@@ -36,14 +36,11 @@
   </publications>
 
   <dependencies>
-    <dependency org="org.apache.any23" name="apache-any23-core" rev="2.2" conf="*->default">
+    <dependency org="org.apache.any23" name="apache-any23-core" rev="2.3" conf="*->default">
       <exclude org="org.apache.commons" name="commons-lang" />
```
Comment (lewismc) [inline src/plugin/any23/ivy.xml:40]: Try adding an entry for 
```
<exclude org="javax.ws.rs" name="javax.ws.rs-api" />
```

Comment (balashashanka) [reply src/plugin/any23/ivy.xml:40]: Let me try this.

Comment (sebastian-nagel) [reply src/plugin/any23/ivy.xml:40]: See also [NUTCH-2669](https://issues.apache.org/jira/browse/NUTCH-2669) and [NUTCH-2697](https://issues.apache.org/jira/browse/NUTCH-2697) - short summary: the issue is fixed with ivy 2.5.0 but the upgrade would require that the local ivy cache is exclusively written by 2.5.0 and not any older ivy version, a requirement hard to ensure on shared dev or nightly build systems. Luckily, Tika dropped the javax.ws.rs-api dependency.

Comment (balashashanka) [reply src/plugin/any23/ivy.xml:41]: After adding the exclude I was able to build it in local.

**Comment by lewismc** [comment]
Hi @balashashanka OK I removed local ant and ivy cache
```
rm -r ~/.ivy2
rm -r ~/.ant
```
I was then able to build the project however now tests are failing. See attached plugin test log 

[TEST-org.apache.nutch.any23.TestAny23ParseFilter.txt](https://github.com/apache/nutch/files/4992711/TEST-org.apache.nutch.any23.TestAny23ParseFilter.txt)

**Comment by lewismc** [comment]
Additionally,
```
./runtime/local/bin/nutch parsechecker http://nutch.apache.org
fetching: http://nutch.apache.org
robots.txt whitelist not configured.
parsing: http://nutch.apache.org
contentType: text/html
signature: 0a441dff545768701b539bf0ea7407bf
http://nutch.apache.org
Version: 5
Status: failed(2,200): org.apache.nutch.parse.ParseException: Unable to successfully parse content
Title:
Outlinks: 0
Content Metadata:
  nutch.crawl.score = 0.0
Parse Metadata:
```
The hadoop.log looks as follows
```
cat runtime/local/logs/hadoop.log
2020-07-31 18:33:05,687 INFO  parse.ParserChecker - fetching: http://nutch.apache.org
2020-07-31 18:33:05,789 INFO  protocol.RobotRulesParser - robots.txt whitelist not configured.
2020-07-31 18:33:05,790 INFO  http.Http - http.proxy.host = null
2020-07-31 18:33:05,790 INFO  http.Http - http.proxy.port = 8080
2020-07-31 18:33:05,790 INFO  http.Http - http.proxy.exception.list = false
2020-07-31 18:33:05,791 INFO  http.Http - http.timeout = 10000
2020-07-31 18:33:05,791 INFO  http.Http - http.content.limit = 1048576
2020-07-31 18:33:05,791 INFO  http.Http - http.agent = lewismc/Nutch-1.18-SNAPSHOT
2020-07-31 18:33:05,791 INFO  http.Http - http.accept.language = en-us,en-gb,en;q=0.7,*;q=0.3
2020-07-31 18:33:05,791 INFO  http.Http - http.accept = text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
2020-07-31 18:33:05,791 INFO  http.Http - http.enable.cookie.header = true
2020-07-31 18:33:06,217 ERROR extractor.ExtractorRegistryImpl - Failed to instantiate service
java.util.ServiceConfigurationError: org.apache.any23.extractor.ExtractorFactory: Provider org.apache.any23.extractor.rdf.FunctionalSyntaxExtractorFactory could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.eclipse.rdf4j.common.lang.service.ServiceRegistry.<init>(ServiceRegistry.java:43)
	at org.apache.any23.extractor.ExtractorRegistryImpl.<init>(ExtractorRegistryImpl.java:45)
	at org.apache.any23.extractor.ExtractorRegistryImpl.getInstance(ExtractorRegistryImpl.java:56)
	at org.apache.any23.Any23.<init>(Any23.java:134)
	at org.apache.any23.Any23.<init>(Any23.java:144)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.parse(Any23ParseFilter.java:101)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.<init>(Any23ParseFilter.java:83)
	at org.apache.nutch.any23.Any23ParseFilter.filter(Any23ParseFilter.java:155)
	at org.apache.nutch.parse.HtmlParseFilters.filter(HtmlParseFilters.java:45)
	at org.apache.nutch.parse.html.HtmlParser.getParse(HtmlParser.java:256)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:34)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:23)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ExceptionInInitializerError
	at org.eclipse.rdf4j.rio.RDFFormat.<clinit>(RDFFormat.java:72)
	at org.apache.any23.extractor.rdf.FunctionalSyntaxExtractorFactory.<init>(FunctionalSyntaxExtractorFactory.java:42)
	at org.apache.any23.extractor.rdf.FunctionalSyntaxExtractorFactory.<clinit>(FunctionalSyntaxExtractorFactory.java:39)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 18 more
Caused by: java.lang.ClassCastException: org.apache.xerces.jaxp.datatype.DatatypeFactoryImpl cannot be cast to javax.xml.datatype.DatatypeFactory
	at javax.xml.datatype.DatatypeFactory.newInstance(Unknown Source)
	at org.eclipse.rdf4j.model.impl.AbstractValueFactory.<clinit>(AbstractValueFactory.java:53)
	... 27 more
2020-07-31 18:33:06,219 ERROR extractor.ExtractorRegistryImpl - Failed to instantiate service
java.util.ServiceConfigurationError: org.apache.any23.extractor.ExtractorFactory: Provider org.apache.any23.extractor.rdf.ManchesterSyntaxExtractorFactory could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.eclipse.rdf4j.common.lang.service.ServiceRegistry.<init>(ServiceRegistry.java:43)
	at org.apache.any23.extractor.ExtractorRegistryImpl.<init>(ExtractorRegistryImpl.java:45)
	at org.apache.any23.extractor.ExtractorRegistryImpl.getInstance(ExtractorRegistryImpl.java:56)
	at org.apache.any23.Any23.<init>(Any23.java:134)
	at org.apache.any23.Any23.<init>(Any23.java:144)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.parse(Any23ParseFilter.java:101)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.<init>(Any23ParseFilter.java:83)
	at org.apache.nutch.any23.Any23ParseFilter.filter(Any23ParseFilter.java:155)
	at org.apache.nutch.parse.HtmlParseFilters.filter(HtmlParseFilters.java:45)
	at org.apache.nutch.parse.html.HtmlParser.getParse(HtmlParser.java:256)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:34)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:23)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.semanticweb.owlapi.rio.OWLAPIRDFFormat
	at org.apache.any23.extractor.rdf.ManchesterSyntaxExtractorFactory.<init>(ManchesterSyntaxExtractorFactory.java:42)
	at org.apache.any23.extractor.rdf.ManchesterSyntaxExtractorFactory.<clinit>(ManchesterSyntaxExtractorFactory.java:39)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 18 more
2020-07-31 18:33:06,220 ERROR extractor.ExtractorRegistryImpl - Failed to instantiate service
java.util.ServiceConfigurationError: org.apache.any23.extractor.ExtractorFactory: Provider org.apache.any23.extractor.rdf.NQuadsExtractorFactory could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.eclipse.rdf4j.common.lang.service.ServiceRegistry.<init>(ServiceRegistry.java:43)
	at org.apache.any23.extractor.ExtractorRegistryImpl.<init>(ExtractorRegistryImpl.java:45)
	at org.apache.any23.extractor.ExtractorRegistryImpl.getInstance(ExtractorRegistryImpl.java:56)
	at org.apache.any23.Any23.<init>(Any23.java:134)
	at org.apache.any23.Any23.<init>(Any23.java:144)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.parse(Any23ParseFilter.java:101)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.<init>(Any23ParseFilter.java:83)
	at org.apache.nutch.any23.Any23ParseFilter.filter(Any23ParseFilter.java:155)
	at org.apache.nutch.parse.HtmlParseFilters.filter(HtmlParseFilters.java:45)
	at org.apache.nutch.parse.html.HtmlParser.getParse(HtmlParser.java:256)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:34)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:23)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.eclipse.rdf4j.rio.RDFFormat
	at org.apache.any23.extractor.rdf.NQuadsExtractorFactory.<init>(NQuadsExtractorFactory.java:42)
	at org.apache.any23.extractor.rdf.NQuadsExtractorFactory.<clinit>(NQuadsExtractorFactory.java:39)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 18 more
2020-07-31 18:33:06,232 WARN  parse.ParseUtil - Error parsing http://nutch.apache.org with org.apache.nutch.parse.html.HtmlParser
java.util.concurrent.ExecutionException: java.lang.ExceptionInInitializerError
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at org.apache.nutch.parse.ParseUtil.runParser(ParseUtil.java:188)
	at org.apache.nutch.parse.ParseUtil.parse(ParseUtil.java:92)
	at org.apache.nutch.parse.ParserChecker.process(ParserChecker.java:260)
	at org.apache.nutch.util.AbstractChecker.processSingle(AbstractChecker.java:87)
	at org.apache.nutch.parse.ParserChecker.run(ParserChecker.java:145)
	at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:76)
	at org.apache.nutch.parse.ParserChecker.main(ParserChecker.java:302)
Caused by: java.lang.ExceptionInInitializerError
	at org.apache.tika.config.TikaConfig.<init>(TikaConfig.java:158)
	at org.apache.any23.mime.TikaMIMETypeDetector.<init>(TikaMIMETypeDetector.java:205)
	at org.apache.any23.Any23.<init>(Any23.java:83)
	at org.apache.any23.Any23.<init>(Any23.java:133)
	at org.apache.any23.Any23.<init>(Any23.java:144)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.parse(Any23ParseFilter.java:101)
	at org.apache.nutch.any23.Any23ParseFilter$Any23Parser.<init>(Any23ParseFilter.java:83)
	at org.apache.nutch.any23.Any23ParseFilter.filter(Any23ParseFilter.java:155)
	at org.apache.nutch.parse.HtmlParseFilters.filter(HtmlParseFilters.java:45)
	at org.apache.nutch.parse.html.HtmlParser.getParse(HtmlParser.java:256)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:34)
	at org.apache.nutch.parse.ParseCallable.call(ParseCallable.java:23)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: org.apache.xerces.jaxp.SAXParserFactoryImpl cannot be cast to javax.xml.parsers.SAXParserFactory
	at javax.xml.parsers.SAXParserFactory.newInstance(Unknown Source)
	at org.apache.tika.utils.XMLReaderUtils.getSAXParserFactory(XMLReaderUtils.java:216)
	at org.apache.tika.utils.XMLReaderUtils.setPoolSize(XMLReaderUtils.java:611)
	at org.apache.tika.utils.XMLReaderUtils.<clinit>(XMLReaderUtils.java:112)
	... 16 more
2020-07-31 18:33:06,232 WARN  parse.ParseUtil - Unable to successfully parse content http://nutch.apache.org of type text/html
2020-07-31 18:33:06,234 INFO  crawl.SignatureFactory - Using Signature impl: org.apache.nutch.crawl.MD5Signature
2020-07-31 18:33:06,238 INFO  parse.ParserChecker - parsing: http://nutch.apache.org
2020-07-31 18:33:06,238 INFO  parse.ParserChecker - contentType: text/html
2020-07-31 18:33:06,238 INFO  parse.ParserChecker - signature: 0a441dff545768701b539bf0ea7407bf
```

**Comment by sebastian-nagel** [comment]
Fyi, the issues around javax.ws.rs-api and ${packaging.type} are now fixed in master by upgrading to Ivy 2.5.0, see NUTCH-2697.

**Comment by lewismc** [comment]
Thanks @sebastian-nagel 

@balashashanka I just pushed a [release candidate for Any23 2.4](https://lists.apache.org/thread.html/r304bca205fa645e7d00522e4d50ad02c0aecdcbf429c00893a0f9f43%40%3Cdev.any23.apache.org%3E). It would be great if you could check it out and provide a VOTE 👍 

Once 2.4 is released, it's most likely worthwhile us closing this off and creating a clean PR against master.

**Comment by balashashanka** [comment]
Hey @lewismc, sure will take a look into it...
And i accept saying once the new version is released I will work on much more cleaner PR.

**Comment by lewismc** [comment]
We just released Any23 2.4 so I'mn going to close and submit a new PR. Thanks @balashashanka for initiating the upgrade

**Pull request closed without merging**